File: sop-java-picocli/src/main/java/sop/cli/picocli/commands/ChangeKeyPasswordCmd.java
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands;

import picocli.CommandLine;
import sop.cli.picocli.SopCLI;
import sop.exception.SOPGPException;
import sop.operation.ChangeKeyPassword;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

@CommandLine.Command(name = "change-key-password",
        resourceBundle = "msg_change-key-password",
        exitCodeOnInvalidInput = SOPGPException.UnsupportedOption.EXIT_CODE)
public class ChangeKeyPasswordCmd extends AbstractSopCmd {

    @CommandLine.Option(names = "--no-armor",
            negatable = true)
    boolean armor = true;

    @CommandLine.Option(names = {"--old-key-password"})
    List<String> oldKeyPasswords = new ArrayList<>();

    @CommandLine.Option(names = {"--new-key-password"}, arity = "0..1")
    String newKeyPassword = null;

    @Override
    public void run() {
        ChangeKeyPassword changeKeyPassword = throwIfUnsupportedSubcommand(
                SopCLI.getSop().changeKeyPassword(), "change-key-password");

        if (!armor) {
            changeKeyPassword.noArmor();
        }

        for (String oldKeyPassword : oldKeyPasswords) {
            changeKeyPassword.oldKeyPassphrase(oldKeyPassword);
        }

        if (newKeyPassword != null) {
            changeKeyPassword.newKeyPassphrase(newKeyPassword);
        }

        try {
            changeKeyPassword.keys(System.in).writeTo(System.out);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: sop-java-picocli/src/main/kotlin/sop/cli/picocli/commands/ChangeKeyPasswordCmd.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package sop.cli.picocli.commands

import java.io.IOException
import java.lang.RuntimeException
import picocli.CommandLine.Command
import picocli.CommandLine.Option
import sop.cli.picocli.SopCLI
import sop.exception.SOPGPException

@Command(
    name = "change-key-password",
    resourceBundle = "msg_change-key-password",
    exitCodeOnInvalidInput = SOPGPException.UnsupportedOption.EXIT_CODE)
class ChangeKeyPasswordCmd : AbstractSopCmd() {

    @Option(names = [OPT_NO_ARMOR], negatable = true) var armor: Boolean = true

    @Option(names = [OPT_OLD_KEY_PASSWORD], paramLabel = "PASSWORD")
    var oldKeyPasswords: List<String> = listOf()

    @Option(names = [OPT_NEW_KEY_PASSWORD], arity = "0..1", paramLabel = "PASSWORD")
    var newKeyPassword: String? = null

    override fun run() {
        val changeKeyPassword =
            throwIfUnsupportedSubcommand(SopCLI.getSop().changeKeyPassword(), "change-key-password")

        if (!armor) {
            throwIfUnsupportedOption(OPT_NO_ARMOR) { changeKeyPassword.noArmor() }
        }

        oldKeyPasswords.forEach {
            val password = stringFromInputStream(getInput(it))
            changeKeyPassword.oldKeyPassphrase(password)
        }

        newKeyPassword?.let {
            val password = stringFromInputStream(getInput(it))
            changeKeyPassword.newKeyPassphrase(password)
        }

        try {
            changeKeyPassword.keys(System.`in`).writeTo(System.out)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    companion object {
        const val OPT_NO_ARMOR = "--no-armor"
        const val OPT_OLD_KEY_PASSWORD = "--old-key-password"
        const val OPT_NEW_KEY_PASSWORD = "--new-key-password"
    }
}
